APE17.rst

```diff
@@ -0,0 +1,482 @@
+
+Therefore, we recommend that Pure-python packages can make all the changes
+described here now. On the other hand, packages that need to rely on
+``pyproject.toml`` for building C extensions should make the switch at the time
+they are happy to rely on pip 19.0 (released 2019-01-22) for installs from
+source distributions, but they can still make other changes now, e.g. using
+``setuptools_scm`` or making a greater use of ``setup.cfg``.
+
+We note however that provided that wheels are available for a package with
+compiled extensions, it may be acceptable to transition to using
+``pyproject.toml`` sooner rather than later because pip 19.0 would only be
+required for source installs, but since most users that pip install would be
+using wheels, this may not be an issue.
+
+Impact for users
+----------------
+
+If done properly, these changes should have no noticeable impact for users.
+Users will still be able to pip install (or conda install when available)
+packages and for large packages such as the core astropy package, which contains
+a lot of compiled extensions, most users will not see any
+difference since they will be installing astropy from pre-built packages (wheels
+or conda packages). Running tests using ``package.test()`` will still be
+supported.
+
+Impact for non-core developers
+------------------------------
+
+The immediate impact for developers of packages that use astropy-helpers is
+having to update the layout and infrastructure in their packages to follow the
+new guidelines presented here. However, we emphasise that these changes are
+recommendations and not mandatory, and astropy-helpers will continue to work as
+expected as long as it is included as a submodule. However, we recommend that
+astropy-helpers no longer be actively developed, in which case guarantees could
+not be made that some aspects of the astropy-helpers infrastructure will not
+break with future releases of Python, setuptools, or sphinx for example.
+
+However, we believe that the initial effort to switch over to the new guidelines
+will be a worthwhile investment - in all cases it will mean being able to get
+rid of astropy-helpers as a submodule and the confusion and headaches this can
+cause, and it will make for example the definition of package data much simpler
+and not have to worry about ``setup_package.py`` files in most cases.
+
+The astropy `package-template`_ will be updated to reflect the latest
+recommendations, which will make it easier for developers to update their
+packages.
+
+Impact for contributors
+-----------------------
+
+Users who wish to contribute fixes to the core astropy package or other packages
+will be encouraged to have `tox`_ installed if they want to easily run
+tests or build documentation locally. However, this is an easy package to
+install with `pip`_ and we could also add code in `setup.py` so that running
+e.g. ``python setup.py test`` or ``python setup.py build_docs`` gives a helpful
```

Comment:
This docs washer over on the fact that the current `python setup.py test` solution provides access to all the conveniences the astropy test runner has. Neither tox nor pytest provides any of those, so this proposal means removal of features many of the core devs relied on (e.g. don't underestimate the importance of e.g. `-P table,units` picks up all the tests from both subpackages and their documentations. It's not just much more typing but more error prone to need to think about where all the relevant tests live in a repo).

Reply:
As mentioned above, we could add this back as a pytest option. What other test runner options do you find useful?

Reply:
`-R` is another such nice conveniences

Reply:
What about the `--remote-data` option? Also, when I develop tests, I like to use the pytest options, which can be used via `python setup.py test -a <pytest args>`, e.g., `-x` to exit after first failure. (Several of my packages have C extensions, so running pytest directly wouldn't be possible.)

Reply:
`-R` is short for `--remote-data` 😉

Reply:
I know, it was just a race condition 😁

Reply:
Ok I've added a section on this, and a link to my PR to pytest-astropy to add the ``-P`` option for pytest: https://github.com/astropy/pytest-astropy/pull/19

Reply:
I've added a table showing that most of the previous options/conveniences translate over to options with pytest: https://github.com/astropy/astropy-APEs/pull/52/commits/8841b8aea62edab63cf19e029c96922cb1915f55